crates/sui-core/src/authority_aggregator.rs

```diff
@@ -962,12 +1073,87 @@ where
             Err(state) => {
                 self.record_process_transaction_metrics(tx_digest, &state);
                 let state = Self::record_non_quorum_effects_maybe(tx_digest, state);
-                Err(QuorumSignTransactionError {
-                    total_stake: self.committee.total_votes,
-                    good_stake: state.good_stake(),
+                Err(self.handle_process_transaction_error(tx_digest, state))
+            }
+        }
+    }
+
+    // TODO make sure this is epoch-boundary-safe
+    fn handle_process_transaction_error(
+        &self,
+        original_tx_digest: &TransactionDigest,
+        state: ProcessTransactionState,
+    ) -> AggregatorProcessTransactionError {
+        let validity_threshold = self.committee.validity_threshold();
+
+        if state.retryable {
+            if !state.conflicting_tx_digests.is_empty() {
+                if let Some((conflicting_tx, validators, conflicting_stake)) =
+                    state.conflicting_tx_digest_with_most_stake()
+                {
+                    let good_stake = state.good_stake();
+                    if good_stake >= validity_threshold && conflicting_stake >= validity_threshold {
+                        warn!(
+                            ?conflicting_tx,
+                            ?original_tx_digest,
+                            original_tx_stake = good_stake,
+                            tx_stake = conflicting_stake,
+                            "Equivocation detected: {:?}",
+                            validators
+                        );
+                        self.metrics.total_equivocation_detected.inc();
+                        return AggregatorProcessTransactionError::FatalConflictingTransaction {
+                            errors: state.errors,
+                            conflicting_tx_digests: state.conflicting_tx_digests,
+                        };
+                    } else if good_stake >= validity_threshold {
+                        // We have >= f+1 good stake on the current transaction so we will retry that,
+                        // no point in retrying conflicting ones.
+                        return AggregatorProcessTransactionError::RetryableConflictingTransaction {
+                            errors: state.errors,
+                            tx_digest_to_retry: None,
+                            conflicting_tx_digests: state.conflicting_tx_digests,
+                        };
+                    } else if conflicting_stake >= validity_threshold {
```

Comment:
Previously this code is in QD, and we stop collecting response in AA immediately after we have f+1 errors. Now this is moved to AA so it's likely we have more information at hand to decide if the conflicting tx is retryable? so it will be like
```
this_promising_conflict_tx_stake >= validity && total_other_conflicting_tx_stake + good_stake + non_retryable < validity
```